contracts/voting-with-uuid/src/contract.rs

```diff
@@ -0,0 +1,516 @@
+        let largest_staked = locked_amount(&sender_address_raw, deps.storage);
+        let withdraw_amount = amount.unwrap_or(token_manager.token_balance);
+        if largest_staked + withdraw_amount > token_manager.token_balance {
+            let max_amount = token_manager.token_balance.checked_sub(largest_staked)?;
+            Err(ContractError::ExcessiveWithdraw { max_amount })
+        } else {
+            let balance = token_manager.token_balance.checked_sub(withdraw_amount)?;
+            token_manager.token_balance = balance;
+
+            bank(deps.storage).save(sender_address_raw, &token_manager)?;
+
+            let mut state = config(deps.storage).load()?;
+            let staked_tokens = state.staked_tokens.checked_sub(withdraw_amount)?;
+            state.staked_tokens = staked_tokens;
+            config(deps.storage).save(&state)?;
+
+            Ok(send_tokens(
+                &info.sender,
+                vec![coin(withdraw_amount.u128(), &state.denom)],
+                "approve",
+            ))
+        }
+    } else {
+        Err(ContractError::PollNoStake {})
+    }
+}
+
+/// validate_description returns an error if the description is invalid
+fn validate_description(description: &str) -> Result<(), ContractError> {
+    if (description.len() as u64) < MIN_DESC_LENGTH {
+        Err(ContractError::DescriptionTooShort {
+            min_desc_length: MIN_DESC_LENGTH,
+        })
+    } else if (description.len() as u64) > MAX_DESC_LENGTH {
+        Err(ContractError::DescriptionTooLong {
+            max_desc_length: MAX_DESC_LENGTH,
+        })
+    } else {
+        Ok(())
+    }
+}
+
+/// validate_quorum_percentage returns an error if the quorum_percentage is invalid
+/// (we require 0-100)
+fn validate_quorum_percentage(quorum_percentage: Option<u8>) -> Result<(), ContractError> {
+    match quorum_percentage {
+        Some(qp) => {
+            if qp > 100 {
+                return Err(ContractError::PollQuorumPercentageMismatch {
+                    quorum_percentage: qp,
+                });
+            }
+            Ok(())
+        }
+        None => Ok(()),
+    }
+}
+
+/// validate_end_height returns an error if the poll ends in the past
+fn validate_end_height(end_height: Option<u64>, env: Env) -> Result<(), ContractError> {
+    if end_height.is_some() && env.block.height >= end_height.unwrap() {
+        Err(ContractError::PollCannotEndInPast {})
+    } else {
+        Ok(())
+    }
+}
+
+/// create a new poll
+pub fn create_poll(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    quorum_percentage: Option<u8>,
+    description: String,
+    start_height: Option<u64>,
+    end_height: Option<u64>,
+) -> Result<Response, ContractError> {
+    validate_quorum_percentage(quorum_percentage)?;
+    validate_end_height(end_height, env.clone())?;
+    validate_description(&description)?;
+
+    let mut state = config(deps.storage).load()?;
```

Comment:
nit: not need to be mutable

Reply:
I left it in the changing to using uuid. Deleted. thanks!